Add vl53l1x support

overheat opened the pull request:
## VL53L1X time of flight (TOF) sensor supporting
https://www.st.com/en/imaging-and-photonics-solutions/vl53l1x.html

### The vl53l1 library
To ease the usage of its vl53l1x time of flight sensor, STMicroelectronics provides this library, which is made of 2 parts :

- the core, that is generic to all platforms
- the platform, contains the adaptation layer, which will be implemented in the driver as it is Zephyr specific.

### The vl53l1x driver
Contains vl53l1x driver itself and the adaptation layer, which is implemented in the driver as it is Zephyr specific.

### The vl53l1x sample
vl53l1x  sensor sample, fetching proximity and distance data in every second. 

### Misc
documentation and overlay file

overheat committed:
ext: Add official ST library for vl53l1x

This package contains platform independent drivers written in C
language for STMicroelectronics ToF sensors VL53L1X. The aim of
this package is to provide a common, clean and stable interface
to get ranging data.

Library is located in ext/hal/st/lib/sensor/vl53l1x

Origin: ST Microelectronics
License: BSD-3-Clause
URL: http://www.st.com/en/embedded-software/stsw-img007.html
Commit: v2.3.3
Purpose: provide a common and stable i/f to get ranging data
Maintained-by: ST Microelectronics

Signed-off-by: Aaron Tsui <aaron.tsui@outlook.com>

dbkinder commented on samples/sensor/vl53l1x/README.rst:
```diff
@@ -0,0 +1,48 @@
+
+Overview
+********
+This sample periodically measures distance between vl53l1x sensor
```
measures the distance between the VL53L1X sensor and a target

dbkinder commented on samples/sensor/vl53l1x/README.rst:
```diff
@@ -0,0 +1,48 @@
+********
+This sample periodically measures distance between vl53l1x sensor
+and target. The result is displayed on the console.
+It also shows how we can use the vl53l1x as a proximity sensor.
```
VL53L1X

overheat replied:
Thank you! Already changed.

MaureenHelm commented on ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h:
```diff
@@ -0,0 +1,236 @@
+* This file is part of VL53L1 Core and is dual licensed,
+* either 'STMicroelectronics
+* Proprietary license'
+* or 'BSD 3-clause "New" or "Revised" License' , at your option.
```
Please review https://docs.zephyrproject.org/latest/contribute/index.html#contributing-non-apache-2-0-licensed-components and fill out the README template.

cc: @avisconti

overheat replied:
@avisconti  Thanks!
I have a question that those code under ext/hal/st/ comes from ST. Can I modify the copyright part as vl53l0x does?

avisconti replied:
@overheat 
For the ext part yes. You may refer to the existing VL53L0X for reference.
For the driver/sensor/vl53l1x you may change/add ur name and 2019 in the copyright section.

I think you may need to change CODEOWNER to cover that code. You should be the owner for the driver, while I think that someone from ST should for what concern the ext part (maybe myself or @erwango)

overheat replied:
@avisconti Thank you! I will update it. I will put you the CODEOWNER for the ext part, is that alright?

avisconti replied:
@overheat 
OK, fine.

avisconti replied:
@overheat 
I'm discussing with @erwango. I think that he can be the owner for it.

avisconti replied:
So, at the end I think that you won't need to change CODEOWNERS at all.

avisconti commented on drivers/sensor/vl53l1x/vl53l1_platform.c:
```diff
@@ -0,0 +1,244 @@
+}
+
+VL53L1_Error VL53L1_ReadMulti(VL53L1_DEV Dev, uint16_t index, uint8_t *pdata,
+			      uint32_t count)
```
Same here.
What is the difference with VL53L0?
index is on 16-bit instead of 8-bit?

overheat replied:
Yes, the VL53L1 index is on 16-bit instead of 8-bit

I have thought about re-use same driver but looks the internal implication is different between vl53l1x and vl53l0x, as you already have seen the "index"(register address) length is not the same. Also, those API related to interrupt and threshold and so on. So there should be lots of ifdef, if we want to merge those drivers together.

avisconti replied:
> I have thought about re-use same driver but looks the internal implication is different between vl53l1x and vl53l0x, as you already have seen the "index"(register address) length is not the same. Also, those API related to interrupt and threshold and so on. So there should be lots of ifdef, if we want to merge those drivers together.

Yes, yes. In fact my initial idea was to re-use same driver, but the more I was going deep thru the review the less I found it applicable. So at the end I left it as a comment from your side.

avisconti reviewed (changes requested):
I'm not sure if it is possible to reuse some code for both VL53L0 and VL53L1 sensor. Nor if it is useful.
Maybe separating them is the right strategy. Pls add some comments to this.

If separation is required you may still want to adapt the copyright date.

overheat commented:
@avisconti @MaureenHelm @dbkinder  I have modified. Please review. Thanks!

avisconti commented:
> @avisconti @MaureenHelm @dbkinder I have modified. Please review. Thanks!

Well, if I'm getting the change I guess this is not what we were expecting.
I mean, you do not have to change the ext code. The ext code should be left there unchanged, with the
original copyright message. At least this is what I have been told when I did a similar change.

You can take as an example the PR #14111 and especially the 52bc149 commit. You will see
the ext code unchanged from the original but with the following description in the commit msg:

commit 52bc149f0787e51a3ee44ad7833fec50cb0e34a1
Author: Armando Visconti <armando.visconti@st.com>
Date:   Tue Mar 5 15:56:18 2019 +0100

    ext/hal: Introduce STMEMS Standard C Drivers
    
    This package contains platform independent drivers written in C
    language for STMicroelectronics sensors. The aim of this package
    is to provide a common, clean and stable interface to access sensor
    registers.
    
    Library is located in ext/hal/st/stmemsc/
    
    Origin: ST Microelectronics
    License: BSD-3-Clause
    URL: https://www.st.com/en/embedded-software/c-driver-mems.html
    Commit: v1.00
    Purpose: provide a common and stable i/f to access sensor registers
    Maintained-by: ST Microelectronics
    
    Signed-off-by: Armando Visconti <armando.visconti@st.com>

Moreover, I also added a new README file: ext/hal/st/stmemsc/README (see again #14111)

So, to summarize, you have to keep the ext code original, then add a README file as I did, then
add the info in the commit msg as I did. Is it clear?

Instead, what I asked you to change is the copyright of the drivers/sensor/vl53l1x driver.
There you need to put 2019 as a date and your name.

Does it make sense to you?

dbkinder reviewed (approved):
Doc changes look OK, thanks.

overheat commented:
@avisconti Thank you for your patience. I have restored the ext code back to  ST code original and add a README file, also modify the info in the commit msg. meanwhile, change the copyright of the drivers/sensor/vl53l1x driver.
Please help to review again, thanks a lot!

avisconti commented on ext/hal/st/lib/sensor/vl53l1x/README:
```diff
@@ -0,0 +1,53 @@
+
+   In order to use this library, you have to :
+      * define CONFIG_HAS_STLIB and CONFIG_VL53L1X
+      * use NEWLIB_LIBC in prj.conf (needed for abs function)
```
I'm currently adding a similar external hal from ST Microelectronics for other sensors. I also need to
define NEWLIB_LIBC, but instead of letting the user to define it in prj.conf I used imply in Kconfig of the external
code. Have you considered to do the same?
See #14111 and in particular a40e62975 commit.

overheat replied:
yes, sound better. 
wait a moment, I will push again.

avisconti replied:
OK, just remember to squash together similar commits as I was saying before.
I suggest to save the branch before squashing just in case

avisconti reviewed (changes requested):
We are almost done. See my latest comment here about using 'imply' in Kconfig for NEWLIB_LIBC.
Moreover, can you squash together the related commits?
At the end I think that you should have only 3 commits:
  - ext: Add official ST library for vl53l1x
  - drivers: sensor: Add vl53l1x time of flight sensor driver
  - samples: sensor: Add vl53l1x time of flight sensor sample

Does it make sense?

overheat committed:
drivers: sensor: Add vl53l1x time of flight sensor driver

Contains vl53l1x driver itself and the adaptation layer, which is
implemented in the driver as it is Zephyr specific.

Support its own thread (minimum latency) or share a system-wide thread
(less memory) to fetch data. Also, need to clear interrupt flag after
every measurement on vl53l1x.

Add dts yaml file as dts/bindings/sensor/st,vl53l1x.yaml

Signed-off-by: Aaron Tsui <aaron.tsui@outlook.com>

overheat committed:
samples: sensor: Add vl53l1x time of flight sensor sample

vl53l1x time of flight (TOF) sensor sample, fetching proximity and
distance data in every second.

Add overlay file for nucleo_f401re to make this sample works out-of-box.

Signed-off-by: Aaron Tsui <aaron.tsui@outlook.com>

avisconti reviewed (approved):
Good job!

overheat commented:
Fix mirror bugs.

avisconti committed:
drivers/i2s_ll_stm32.c: (FIX) Avoid warning in LOG_ERR

The LL_I2S_ReadReg() function returns uint32_t, while %d requires
'int' as a type.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@overheat 
Pls take a look at #15857. It should fix the CI error you are having here.

overheat commented:
>@overheat  ...

@avisconti yes, should I wait until that PR merged then re-trigger this CI check again?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

avisconti commented:
> @avisconti yes, should I wait until that PR merged then re-trigger this CI check again?

Maybe you can get my commit (c18747f) from my remote repo on github, rebase on top of it and re-push everything:

-----
git add remote avisconti 	https://github.com/avisconti/zephyr.git
git fetch avisconti
git cherry-pick c18747f
git push -f ....
-----

I think it is important to know if it is fixing or not.

overheat commented:
> > @avisconti yes, should I wait until that PR merged then re-trigger this CI check again?
> 
> Maybe you can get my commit ([c18747f](https://github.com/zephyrproject-rtos/zephyr/commit/c18747faf548d8743e2f8d6d3d62cdefbca39a16)) from my remote repo on github, rebase on top of it and re-push everything:
> 
> git add remote avisconti https://github.com/avisconti/zephyr.git
> git fetch avisconti
> git cherry-pick [c18747f](https://github.com/zephyrproject-rtos/zephyr/commit/c18747faf548d8743e2f8d6d3d62cdefbca39a16)
> git push -f ....
> 
> I think it is important to know if it is fixing or not.

Done, let's see what happen.

overheat commented:
@avisconti it works. I will push again when #15857 be merged.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

avisconti commented:
superseeded by #18772
